### PR TITLE
Signaling-only CI job to detect mismatching build artifact hashes early

### LINKED ISSUE #1203: Notification for mismatching builds
In #1201 it turns out that @0xB10C had been building mismatching builds for a while, without it being noticed before.

It would be useful to have some kind of notification when this happens. A mismatching build always points to a problem, either (in the good case) a local configuration issue, but it can also be a bug in the build system or worst-case even a backdoor attempt. So it needs to be investigated.

See discussion here: https://github.com/bitcoin-core/guix.sigs/pull/1201#issuecomment-2071818604

### PR DESCRIPTION
**Pull request by 0xB10C**
To detect build non-determinism and mismatching artifact hashes early (https://github.com/bitcoin-core/guix.sigs/issues/1203), this introduces a CI job that comments a visual summary of the SHASUMs for each added/changed release. Before merging, it can be quickly visually verified that there are no mismatches. The CI job does **not** fail on mismatches - it only signals that there's a problem. It also does **not** check signature validity.

For example, when a `noncodesigned.SHA256SUMS` file for `24.2` is added, the CI job would comment on the PR:

----

`noncodesigned.SHA256SUMS` summary for release `24.2`
```
User (see mapping below)            a b c d e f g h i j k l m n o p q
aarch64-linux-gnu-debug.tar.gz      █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
aarch64-linux-gnu.tar.gz            █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
arm-linux-gnueabihf-debug.tar.gz    █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
arm-linux-gnueabihf.tar.gz          █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
arm64-apple-darwin-unsigned.dmg     █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
arm64-apple-darwin-unsigned.tar.gz  █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
arm64-apple-darwin.tar.gz           █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
powerpc64-linux-gnu-debug.tar.gz    █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
powerpc64-linux-gnu.tar.gz          █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
powerpc64le-linux-gnu-debug.tar.gz  █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
powerpc64le-linux-gnu.tar.gz        █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
riscv64-linux-gnu-debug.tar.gz      █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
riscv64-linux-gnu.tar.gz            █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
win64-debug.zip                     █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
win64-setup-unsigned.exe            █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
win64-unsigned.tar.gz               █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
win64.zip                           █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
x86_64-apple-darwin-unsigned.dmg    █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
x86_64-apple-darwin-unsigned.tar.gz █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
x86_64-apple-darwin.tar.gz          █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
x86_64-linux-gnu-debug.tar.gz       █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
x86_64-linux-gnu.tar.gz             █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
bitcoin-24.2.tar.gz                 █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █ █
```

<details><summary>Details</summary>
Symbols:
 - all hashes match: `█`
 - missing hash: `X`
 - hash mismatch: one of `░`, `▞`, `▄`, `▀`, `▌`, `▐`, `▚`

Username mapping:
- a: willyko
- b: jackielove4u
- c: hebasto
- d: kvaciral
- e: TheCharlatan
- f: 0xb10c
- g: laanwj
- h: svanstaa
- i: glozow
- j: Emzy
- k: achow101
- l: Sjors
- m: fanquake
- n: guggero
- o: theStack
- p: benthecarman
- q: josibake

</details>

----

Here, all rows having only `█`'s means that there are **no** hash mismatches. Other examples can be found here: https://github.com/0xB10C/guix.sigs/pull/2 

closes #1203

### CODE REVIEW

**Comment by laanwj** [comment]
i really like the overview, concept ACK!

**Comment by achow101** [comment]
Neat, Concept ACK

**Comment by Mmgg002** [comment]
3fbbc23e4cba0e583344f9abe5348df6ab7d48b7

File: contrib/shasum-summary/main.py
Comment (0xB10C) [reply contrib/shasum-summary/main.py]: done

File: .github/workflows/shasum-summary.yml
Comment (0xB10C) [reply .github/workflows/shasum-summary.yml:1]: done, thanks!

File: .github/workflows/shasum-summary.yml
Comment (0xB10C) [reply .github/workflows/shasum-summary.yml:31]: I didn't have to configure the token when testing on my fork.

`actions/github-script` uses the `GITHUB_TOKEN` by default, which is available in GitHub Actions: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

**Review by TyG34 — APPROVED** [review]

**Comment by achow101** [comment]
ACK dcb80c3eacc8822b9fc91e9ab301bee42d440150

**Comment by 0xB10C** [comment]
On 36bf7c80241f8a56255cdeb7f2e839d924150337 the output looks like this for `27.1rc1`:

`noncodesigned.SHA256SUMS` summary for release `27.1rc1`
```
User (see mapping below)            a b c d e f
aarch64-linux-gnu-debug.tar.gz      █ █ █ █ █ █
aarch64-linux-gnu.tar.gz            █ █ █ █ █ █
arm-linux-gnueabihf-debug.tar.gz    █ █ █ █ █ █
arm-linux-gnueabihf.tar.gz          █ █ █ █ █ █
arm64-apple-darwin-unsigned.tar.gz  █ █ █ █ █ █
arm64-apple-darwin-unsigned.zip     █ █ █ █ █ █
arm64-apple-darwin.tar.gz           █ █ █ █ █ █
powerpc64-linux-gnu-debug.tar.gz    █ █ █ █ █ █
powerpc64-linux-gnu.tar.gz          █ █ █ █ █ █
powerpc64le-linux-gnu-debug.tar.gz  █ █ █ █ █ █
powerpc64le-linux-gnu.tar.gz        █ █ █ █ █ █
riscv64-linux-gnu-debug.tar.gz      █ █ █ █ █ █
riscv64-linux-gnu.tar.gz            █ █ █ █ █ █
win64-debug.zip                     █ █ █ █ █ █
win64-setup-unsigned.exe            █ █ █ █ █ █
win64-unsigned.tar.gz               █ █ █ █ █ █
win64.zip                           █ █ █ █ █ █
x86_64-apple-darwin-unsigned.tar.gz █ █ █ █ █ █
x86_64-apple-darwin-unsigned.zip    █ █ █ █ █ █
x86_64-apple-darwin.tar.gz          █ █ █ █ █ █
x86_64-linux-gnu-debug.tar.gz       █ █ █ █ █ █
x86_64-linux-gnu.tar.gz             █ █ █ █ █ █
bitcoin-27.1rc1.tar.gz              █ █ █ █ █ █
```

<details><summary>Details</summary>
Symbols:

- all hashes match: `█`
- missing hash: `X`
- hash mismatch: one of `◆`, `▞`, `▄`, `▀`, `▌`, `▐`, `▚`

Username mapping:
- a: guggero
- b: achow101
- c: kvaciral
- d: CoinForensics
- e: Emzy
- f: fanquake

</details>

**Comment by laanwj** [comment]
Concept and code review ACK 090b0fd17e8f8994aab684fbd544a335930928db

**Comment by 0xB10C** [comment]
@willcl-ark  and I have been looking into why there aren't comments in #1227 #1228:

- apparently the workflow has to be approved first - we think it's only the first run per contributor - but we can't see if the run requires approval
- there's also a fetch-depth issue (can be fixed by setting the fetch depth to 0 or `number of commits + 1`) https://github.com/0xB10C/guix.sigs/pull/4
- from our testing on my fork, it seems there's a permission issue which doesn't allow a GHA job for a PR by will to post a comment to my repo https://github.com/0xB10C/guix.sigs/pull/5, while my PR can posts a comment https://github.com/0xB10C/guix.sigs/pull/6 (I think there's a fix too)

**Comment by 0xB10C** [comment]
> * from our testing on my fork, it seems there's a permission issue which doesn't allow a GHA job for a PR by will to post a comment to my repo [fetch depth 0 0xB10C/guix.sigs#5](https://github.com/0xB10C/guix.sigs/pull/5), while my PR can posts a comment [Add attestations by willcl-ark for 27.1rc1 non-codesigned 0xB10C/guix.sigs#6](https://github.com/0xB10C/guix.sigs/pull/6) (I think there's a fix too)

There is a of making it work but requires to split up the job into two jobs to restrict external CI runs from exposing the GitHub tokens (e.g. by printing them): One job that generates the comment & and uploads it as an build artifact and a second one that downloads the comment and posts it. See an example here: https://github.com/willcl-ark/guix.sigs/pull/10#issuecomment-2143076500

**Comment by willcl-ark** [comment]
For reference, the two workflows end up looking like this: https://github.com/willcl-ark/guix.sigs/tree/split-workflow/.github/workflows

The issue is that a run triggered by `on: pull_request`, doesn't have a `github_token` with comment permissions. It may be possible to do with `on: pull_request_target` but this risks pull requests from forks doing [bad things](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/).

To read data from a different workflow run, the comment workflow (via `on: workflow`, which _does_ natively have comenting permissions on its `github_token`) requires a Personal Access Token to be saved to _this_ repo as an `Actions>Repository Secret` called `GH_PAT`. It needs very minimal permissions, only `actions: read_only`:

image

It's unclear to me if this PAT must be from an owner of the repo, or can be someone else with sufficient permissions affiliated from the repo.

**Comment by achow101** [comment]
Oops forgot to enable actions for this repo. Runs [now](https://github.com/bitcoin-core/guix.sigs/actions/runs/9470725228/job/26092297387), but fetch depth needs to be increased.

**Comment by 0xB10C** [comment]
Fixed in https://github.com/bitcoin-core/guix.sigs/pull/1263

We found out that we can use the `GITHUB_TOKEN` to fetch the artifacts instead of requiring an extra `GH_PAT` in the `on: workflow` workflow. The fine-grained `GH_PAT` with `actions: read_only` would have to be renewed every year - that just adds friction and more work for maintainers.

**Pull request merged**
